### PR TITLE
[AArch64] Extend vecreduce to udot/sdot transformation to support usdot

### PR DESCRIPTION
**Pull request by igogo-x86**

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/neon-dotreduce.ll [hunk 1]
```diff
@@ -860,19 +1213,253 @@ entry:
   ret i32 %x
 }
 
-define i32 @test_udot_v24i8(ptr nocapture readonly %a, ptr nocapture readonly %b, i32 %sum) {
-; CHECK-SD-LABEL: test_udot_v24i8:
+
```
Comment (igogo-x86) [inline llvm/test/CodeGen/AArch64/neon-dotreduce.ll:1216]: I didn't change the original tests; that's `git diff` doing a poor job. Running `git diff --minimal` gives better representation

File: llvm/test/CodeGen/AArch64/neon-dotreduce.ll [hunk 2]
```diff
@@ -1,22 +1,28 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
-; RUN: llc -mtriple aarch64-linux-gnu -mattr=+dotprod    < %s | FileCheck %s --check-prefixes=CHECK,CHECK-SD
-; RUN: llc -mtriple aarch64-linux-gnu -mattr=+dotprod -global-isel -global-isel-abort=2 < %s 2>&1 | FileCheck %s --check-prefixes=CHECK,CHECK-GI
+; RUN: llc -mtriple aarch64-linux-gnu -mattr=+dotprod,+i8mm    < %s | FileCheck %s --check-prefixes=CHECK,CHECK-SD
+; RUN: llc -mtriple aarch64-linux-gnu -mattr=+dotprod,+i8mm -global-isel -global-isel-abort=2 < %s 2>&1 | FileCheck %s --check-prefixes=CHECK,CHECK-GI
```
Comment (igogo-x86) [inline llvm/test/CodeGen/AArch64/neon-dotreduce.ll:3]: I am not sure it is a good place for the tests, but this file is responsible for testing the functionality of the `performVecReduceAddCombine` function and other tests with `+i8mm` have different purposes.

File: llvm/test/CodeGen/AArch64/neon-dotreduce.ll [hunk 3]
```diff
@@ -2866,6 +4497,362 @@ entry:
   %x = add i32 %r1, %r2
   ret i32 %x
 }
+
+define i32 @test_usdot_v33i8(ptr nocapture readonly %a, ptr nocapture readonly %b, i32 %sum) {
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/neon-dotreduce.ll:4501]: Do we need all these non-standard vector types in the tests?

Comment (igogo-x86) [reply llvm/test/CodeGen/AArch64/neon-dotreduce.ll:4501]: Probably not, but should the same rule be applied to duplicates of `dot`/`sdot`? Which test do you think would be enough to keep?

Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/neon-dotreduce.ll:4501]: There are quite a few of them! So long as there are tests for the usdot (including commuted patterns), then it should be OK. You can at least drop all the test that are not power-of-2 sizes - v5/v33/v24/v25.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**